vue/sbc-common-components/src/services/token.services.ts

```diff
@@ -0,0 +1,119 @@
+      this.kc = Keycloak(keyCloakConfigurl)
+      this.kc.init(this.kcOptions
+      ).success(function (authenticated) {
+        console.info('[TokenServices] is User Authenticated:Syncing' + authenticated)
+        self.syncSessionStorage()
+        resolve(self.kc.token)
+      }).error(function (err) {
+        console.info('[TokenServices] Fatal Error:Cudnt Initialise KC instance' + err)
+        reject(new Error('Cudnt Initialise KC'))
+      })
+    })
+  }
+
+  scheduleRefreshTimer () {
+    console.info('[TokenServices Starting the timer] ')
+    this.scheduleRefreshToken()
+  }
+
+  refreshToken () {
+    console.log('[TokenServices] One time Token Refreshing ')
+    return new Promise((resolve, reject) => {
+      this.kc.updateToken(-1)
+        .success(refreshed => {
+          if (refreshed) {
+            console.log('[TokenServices] One time Token Refreshed ')
+            this.syncSessionStorage()
+            resolve()
+          }
+        })
+        .error(() => {
+          // this.cleanupSession()
+          reject(new Error('Cudnt refresh Token'))
+        })
+    })
+  }
+
+  stopRefreshTimer () {
+    console.info('[TokenServices Stopping the timer] ')
+    clearTimeout(this.timerId)
+  }
+
+  private scheduleRefreshToken () {
+    let self = this
+
+    // check if refresh token is still valid . Or else clear all timers and throw errors
+    let refreshTokenExpiresIn = self.kc.refreshToken['exp'] - Math.ceil(new Date().getTime() / 1000) + self.kc.timeSkew
+    if (refreshTokenExpiresIn < 0) {
+      throw new Error('Refresh Token Expired..No more token refreshes')
+    }
+
+    let expiresIn = self.kc.tokenParsed['exp'] - Math.ceil(new Date().getTime() / 1000) + self.kc.timeSkew
+    let refreshInMilliSeconds = (expiresIn * 1000) - 2000 // refresh before two second of expiry
+    console.info('[TokenServices] Token Expires in %s Seconds:', expiresIn)
+    console.info('[TokenServices] Token Refreshal Scheduled in %s Seconds', (refreshInMilliSeconds / 1000))
+    this.timerId = setTimeout(() => {
+      console.log('[TokenServices] Refreshing Token Attempt: %s ', ++this.counter)
+      this.kc.updateToken(-1)
+        .success(refreshed => {
+          if (refreshed) {
+            console.log('successfully refreshed')
+            this.syncSessionStorage()
+            this.scheduleRefreshToken()
+          }
+        })
+        .error(() => {
+          clearTimeout(this.timerId)
+          console.log('refresh failed')
+        })
+    }, refreshInMilliSeconds)
+    console.log('[TokenServices] Timer id:' + this.timerId)
+  }
+
+  private syncSessionStorage () {
+    sessionStorage.setItem('KEYCLOAK_TOKEN', this.kc.token)
+    sessionStorage.setItem('KEYCLOAK_REFRESH_TOKEN', this.kc.refreshToken)
+    sessionStorage.setItem('ID_TOKEN', this.kc.idToken)
+  }
+
+  decodeToken () {
+    try {
+      let token = sessionStorage.getItem('KEYCLOAK_TOKEN')
+      const base64Url = token.split('.')[1]
```

Comment:
Should check that token != null before splitting it or this will crash. Are you guaranteed that session storage has the token?

Reply:
Thanks.Will add it

Reply:
the whole tokenservice is invalid when the session storage doesnt have token.. so may be I will add a check on init to address that